### PR TITLE
BE | Create `PuzzlesController#index` Endpoint

### PR DESCRIPTION
**Pull request by kem247**
## Changes: 
-  added a route, controller method, and a generated test
- 

---
This is related to # 3

This closes #

[] I linted my work. (RuboCop)

[] I've updated documentation & README. (if necessary)

---
(For Fun!) Please include a link to a meme/gif of your feelings about this branch!

Link: 

Notes:  The user object always validates a zipcode presence, therefore there will never be a user without a zipcode. Either we take out that validation, or frontend creates an error message when someone who is not logged in tries to enter a zipcode. 

Didn;t add tests yet due to wanting confirmation on this issue.

### CODE REVIEW

**Review by MelTravelz — COMMENTED** [review]
Hi Kemi! 

This is a great start with the basic logic in place. Here is what we need to do before this PR can be merged:

1) Please pull down main into your branch on your local computer and handle any merge conflicts there. 
    - This is because I added the necessary items for deploying to Heroku and CI/CD yesterday...which is why this PR did not pass the tests and cannot be merged

2) We should schedule a time to meet so we can review what is necessary for creating an API style app. You're work here has all the basic elements but is not complete. 
    - To get a head start on this, you could review my work in the PR called `BE | Create Users#show Endpoint` which will give you a better idea of what needs to be done.

**Comment by kem247** [comment]
> Hi Kemi!
> 
> This is a great start with the basic logic in place. Here is what we need to do before this PR can be merged:
> 
> 1. Please pull down main into your branch on your local computer and handle any merge conflicts there.
>    
>    * This is because I added the necessary items for deploying to Heroku and CI/CD yesterday...which is why this PR did not pass the tests and cannot be merged
> 2. We should schedule a time to meet so we can review what is necessary for creating an API style app. You're work here has all the basic elements but is not complete.
>    
>    * To get a head start on this, you could review my work in the PR called `BE | Create Faker Data & Users#show Endpoint for BE team to view` which will give you a better idea of what needs to be done.

Hey Mel, thanks for the response. I'll take a look at your previous PRs and we can discuss later about the endpoints. It still doesn't answer my concern 'The user object always validates a zipcode presence, therefore there will never be a user without a zipcode. Either we take out that validation, or frontend creates an error message when someone who is not logged in tries to enter a zipcode'. Also, I see that there is no merge conflicts but tests are failing.

**Comment by MelTravelz** [comment]
- Yes, we'll get the GitHubt Actions tests to pass today, if you need my help I'm here all day.

- I hear your concern about the zip_code:
    - When a user is created and saved in our database, we require a user to have a zip_code because that's how we know where their puzzles are located. So this model validation must stay and must be required.
   - And so, any one who is not logged in can still search for puzzles by any zip_code, they will just not have a button to "request/borrow" the puzzle and they will not see the name of the owner of the puzzle.
   - After the log in, a user can return to that page and again search for puzzles by their own zip code or others.

We can discuss this more at our meeting today! : )

File: app/controllers/api/v1/puzzles_controller.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+      zip_code = params[:zip_code]
+
+      users = User.where(zip_code: zip_code) if zip_code.present?
+      
+      if users != []
+        puzzles = Puzzle.where(user_id: users.pluck(:id))
+        render json: PuzzleSerializer.new(puzzles)
+      elsif puzzles == [] || users = []
+        render json: { error: "Puzzles not found in this area" }, status: "404"
```
Comment (MelTravelz) [inline app/controllers/api/v1/puzzles_controller.rb:13]: This error handling works for now to meet MVP within our time limitation. But in a refactor we'll want to get this working through the ErrorSerializer. : )

File: app/controllers/api/v1/puzzles_controller.rb [hunk 2]
```diff
@@ -0,0 +1,17 @@
+class Api::V1::PuzzlesController < ApplicationController
+
+  def index
+
+      zip_code = params[:zip_code]
+
+      users = User.where(zip_code: zip_code) if zip_code.present?
```
Comment (MelTravelz) [inline app/controllers/api/v1/puzzles_controller.rb:7]: This logic should probably be a method in the models and only called on here in the controller. But again, to meet MVP within out time limits let's just roll with it for now. Nice work.

File: spec/requests/api/v1/puzzles_request_spec.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        expect(parsed_data[:data][0][:attributes].keys).to eq([:user_id, :status, :title, :description, :total_pieces, :notes, :puzzle_image_url])
+        expect(parsed_data[:data][0][:attributes][:user_id]).to eq(puzzle_1.user_id)
+        expect(parsed_data[:data][0][:attributes][:status]).to eq(puzzle_1.status)
+        expect(parsed_data[:data][0][:attributes][:title]).to eq(puzzle_1.title) 
+        expect(parsed_data[:data][0][:attributes][:description]).to eq(puzzle_1.description)
+        expect(parsed_data[:data][0][:attributes][:total_pieces]).to eq(puzzle_1.total_pieces)
+        expect(parsed_data[:data][0][:attributes][:notes]).to eq(puzzle_1.notes)
+        expect(parsed_data[:data][0][:attributes][:puzzle_image_url]).to eq(puzzle_1.puzzle_image_url)
+
+        # NOTE Could refactor tests to not see a puzzle_4 in the test because it's in a diff zipcode
```
Comment (MelTravelz) [inline spec/requests/api/v1/puzzles_request_spec.rb:42]: Excellent note here. And I agree that there should be a test to confirm no "stray" puzzles are found in a call for a specific zip_code.

File: spec/requests/api/v1/puzzles_request_spec.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        headers = { 'CONTENT_TYPE' => 'application/json' }
+        put "/api/v1/puzzles", headers: headers, params: JSON.generate(zipcode_params)
+
+        expect(response).to have_http_status(404)
+
+        parsed_error_data = JSON.parse(response.body, symbolize_names: true)
+        
+        expect(parsed_error_data).to be_a(Hash)
+        expect(parsed_error_data.keys).to eq([:error])
+        expect(parsed_error_data[:error]).to eq("Puzzles not found in this area")
```
Comment (MelTravelz) [inline spec/requests/api/v1/puzzles_request_spec.rb:79]: And these tests would change once we use the ErrorSerializer in a refactor.

**Review by MelTravelz — APPROVED** [review]
This is great work Kemi! I'm impressed with all you've done having only worked with Ruby a little bit in your past. I made some notes for later if we have time for a refactor, but I say let's go ahead and merge this into main!

**Pull request merged**